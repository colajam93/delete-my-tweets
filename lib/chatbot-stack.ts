import * as chatbot from 'aws-cdk-lib/aws-chatbot';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export interface ChatbotStackProps extends cdk.StackProps {
  slackChannelConfigurationName: string;
  slackChannelId: string;
  slackWorkspaceId: string;
  functions?: ReadonlyArray<lambda.IFunction>;
}

export class ChatbotStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ChatbotStackProps) {
    super(scope, id, props);

    const config = new chatbot.SlackChannelConfiguration(this, 'Chatbot', {
      ...props,
      logRetention: logs.RetentionDays.ONE_DAY,
    });
    for (const func of props.functions ?? []) {
      func.grantInvoke(config);
    }
  }
}
