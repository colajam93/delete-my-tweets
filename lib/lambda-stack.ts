import * as lambda_python from '@aws-cdk/aws-lambda-python-alpha';
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

export interface LambdaStackProps extends cdk.StackProps {
  readonly userName: string;
}

export class LambdaStack extends cdk.Stack {
  readonly function: lambda.IFunction;

  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    const ssmKeyConsumerKey = `/app/lambda/delete-my-tweets/${props.userName}/consumer-key`;
    const ssmKeyConsumerSecret = `/app/lambda/delete-my-tweets/${props.userName}/consumer-secret`;
    const ssmKeyAccessToken = `/app/lambda/delete-my-tweets/${props.userName}/access-token`;
    const ssmKeyAccessTokenSecret = `/app/lambda/delete-my-tweets/${props.userName}/access-token-secret`;

    const func = new lambda_python.PythonFunction(this, 'Lambda', {
      entry: 'lambda/',
      index: 'main.py',
      runtime: lambda.Runtime.PYTHON_3_9,
      functionName: `delete-my-tweets-lambda-${props.userName}`,
      logRetention: logs.RetentionDays.ONE_MONTH,
      timeout: cdk.Duration.minutes(15),
      environment: {
        SSM_KEY_CONSUMER_KEY: ssmKeyConsumerKey,
        SSM_KEY_CONSUMER_SECRET: ssmKeyConsumerSecret,
        SSM_KEY_ACCESS_TOKEN: ssmKeyAccessToken,
        SSM_KEY_ACCESS_TOKEN_SECRET: ssmKeyAccessTokenSecret,
      },
    });
    this.function = func;

    for (const key of [ssmKeyConsumerKey, ssmKeyConsumerSecret, ssmKeyAccessToken, ssmKeyAccessTokenSecret]) {
      const k = key.split('/').pop();
      if (k) {
        const param = ssm.StringParameter.fromSecureStringParameterAttributes(this, k, {
          parameterName: key,
        });
        param.grantRead(func);
      }
    }
  }
}
