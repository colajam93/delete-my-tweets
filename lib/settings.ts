import * as fs from 'fs';
import * as yaml from 'js-yaml';

export interface Setting {
  userName: string;
}

export function loadSetting(path: string): Setting {
  return yaml.load(fs.readFileSync(path, 'utf8')) as Setting;
}
