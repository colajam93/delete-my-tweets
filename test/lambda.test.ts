import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaStack } from '../lib/lambda-stack';
import { Setting } from '../lib/settings';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/lambda-stack.ts
test('delete-my-tweets snapshot test', () => {
  const app = new cdk.App();
  const env = {
    region: 'dummy-dummy-1',
    account: 'dummy',
  };
  const setting: Setting = {
    userName: 'user_name',
  };
  const stack = new LambdaStack(app, 'delete-my-tweets-lambda', {
    ...setting,
    env,
  });
  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});
