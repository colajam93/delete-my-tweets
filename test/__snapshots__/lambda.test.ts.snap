// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-my-tweets snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LambdaD247545B": Object {
      "DependsOn": Array [
        "LambdaServiceRoleDefaultPolicyDAE46E21",
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-dummy-dummy-dummy-1",
          "S3Key": "78ecf877db8d34031077b959750b2cfea2ec78124465de66f542d0a241f6d6c8.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "SSM_KEY_ACCESS_TOKEN": "/app/lambda/delete-my-tweets/user_name/access-token",
            "SSM_KEY_ACCESS_TOKEN_SECRET": "/app/lambda/delete-my-tweets/user_name/access-token-secret",
            "SSM_KEY_CONSUMER_KEY": "/app/lambda/delete-my-tweets/user_name/consumer-key",
            "SSM_KEY_CONSUMER_SECRET": "/app/lambda/delete-my-tweets/user_name/consumer-secret",
          },
        },
        "FunctionName": "delete-my-tweets-lambda",
        "Handler": "main.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaLogRetentionDBDA6BEA": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "LambdaD247545B",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaServiceRoleA8ED4D3B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaServiceRoleDefaultPolicyDAE46E21": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:dummy-dummy-1:dummy:parameter/app/lambda/delete-my-tweets/user_name/consumer-key",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:dummy-dummy-1:dummy:parameter/app/lambda/delete-my-tweets/user_name/consumer-secret",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:dummy-dummy-1:dummy:parameter/app/lambda/delete-my-tweets/user_name/access-token",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:dummy-dummy-1:dummy:parameter/app/lambda/delete-my-tweets/user_name/access-token-secret",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaServiceRoleDefaultPolicyDAE46E21",
        "Roles": Array [
          Object {
            "Ref": "LambdaServiceRoleA8ED4D3B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-dummy-dummy-dummy-1",
          "S3Key": "6c0316fef24d0df8a9a705c77052001217d864f49af386539d01df54618cd131.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
